source-chargebee-native/source_chargebee_native/api.py

```diff
@@ -62,22 +64,65 @@ async def _fetch_resource_data(
             params[f"{cursor_field}[before]"] = int(end_date.timestamp())
 
     if start_date and end_date and start_date == end_date:
-        raise ValueError("Start and end dates cannot be the same", {"start_date": start_date, "end_date": end_date})
+        raise ValueError(
+            "Start and end dates cannot be the same",
+            {"start_date": start_date, "end_date": end_date},
+        )
 
     if include_deleted:
         params["include_deleted"] = "true"
 
     if filter_params:
         params.update(filter_params)
 
-    response = APIResponse[resource_type].model_validate_json(
-        await http.request(log, url, params=params)
-    )
-    
-    if not response.list:
-        return [], response.next_offset
+    async def _fetch_data(
+        http: HTTPSession,
+        log: Logger,
+        url: str,
+        params: dict[str, int | str | list[int]],
+        resource_type: type[ChargebeeResourceType],
+    ) -> tuple[list[ChargebeeResourceType], str | None]:
+        response = APIResponse[resource_type].model_validate_json(
+            await http.request(log, url, params=params)
+        )
+
+        if not response.list:
+            return [], response.next_offset
+
+        return response.list, response.next_offset
+
+    max_retries = 10
+    retry_count = 0
+
+    while retry_count < max_retries:
```

Comment:
I would change this condition to `while True`, and then remove the `return [], None` at the end of this function. I was a little confused at first how we could ever get to the final `return [], None` - and indeed it appears we can't, so this will make it a little more obvious about what is happening with the retry logic.